File: src/main/java/com/jcloisterzone/game/phase/ShepherdPlacementConfirmPhase.java
package com.jcloisterzone.game.phase;

import com.jcloisterzone.action.ShepherdPlacementConfirmAction;
import com.jcloisterzone.event.MeepleDeployed;
import com.jcloisterzone.figure.Shepherd;
import com.jcloisterzone.game.state.ActionsState;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.io.message.ShepherdPlacementConfirmMessage;
import com.jcloisterzone.random.RandomGenerator;

public class ShepherdPlacementConfirmPhase extends Phase {
    
	public ShepherdPlacementConfirmPhase(RandomGenerator random, Phase defaultNext, RewindActionContainer rewindActionContainer) {
		super(random, defaultNext, rewindActionContainer);
	}

	@Override
    public StepResult enter(GameState state) {
        // Check if a shepherd was just placed in this turn part
        boolean shepherdPlaced = !state.getCurrentTurnPartEvents()
            .filter(ev -> ev instanceof MeepleDeployed)
            .map(ev -> (MeepleDeployed) ev)
            .filter(ev -> ev.getMeeple() instanceof Shepherd && ev.getMovedFrom() == null).isEmpty();
        
        if (!shepherdPlaced) {
            return next(state);
        }

        return promote(state.setPlayerActions(new ActionsState(state.getTurnPlayer(), new ShepherdPlacementConfirmAction(), false)));
    }
    
    @PhaseMessageHandler
    public StepResult handleShepherdPlacementConfirm(GameState state, ShepherdPlacementConfirmMessage msg) {
        state = clearActions(state);
        return next(state);
    }
}

File: src/main/ts/com/jcloisterzone/game/phase/ShepherdPlacementConfirmPhase.ts
import { MeepleDeployed } from "../../event/MeepleDeployed.js";
import { Shepherd } from "../../figure/Shepherd.js";
import { ShepherdPlacementConfirmAction } from "../../action/ShepherdPlacementConfirmAction.js";
import { ShepherdPlacementConfirmMessage } from "../../io/message/ShepherdPlacementConfirmMessage.js";
import type { PlayerAction } from "../../action/PlayerAction.js";
import { ActionsState } from "../state/ActionsState.js";
import type { GameState } from "../state/GameState.js";
import { Phase, type PhaseHandler } from "./Phase.js";
import type { StepResult } from "./StepResult.js";

/** Confirms a just-placed shepherd before the flock is resolved. */
export class ShepherdPlacementConfirmPhase extends Phase {
  static readonly simpleName = "ShepherdPlacementConfirmPhase";

  enter(state: GameState): StepResult {
    const shepherdPlaced = !state
      .getCurrentTurnPartEvents()
      .filter((ev) => ev instanceof MeepleDeployed)
      .filter(
        (ev) =>
          (ev as MeepleDeployed).getMeeple() instanceof Shepherd &&
          (ev as MeepleDeployed).getMovedFrom() === null,
      )
      .isEmpty();

    if (!shepherdPlaced) {
      return this.next(state);
    }

    return this.promote(
      state.setPlayerActions(
        new ActionsState(
          state.getTurnPlayer()!,
          new ShepherdPlacementConfirmAction() as unknown as PlayerAction<unknown>,
          false,
        ),
      ),
    );
  }

  handleShepherdPlacementConfirm(state: GameState, _msg: ShepherdPlacementConfirmMessage): StepResult {
    state = this.clearActions(state);
    return this.next(state);
  }

  protected override messageHandlers(): Map<Function, PhaseHandler> {
    const m = super.messageHandlers();
    m.set(ShepherdPlacementConfirmMessage, this.handleShepherdPlacementConfirm);
    return m;
  }
}
